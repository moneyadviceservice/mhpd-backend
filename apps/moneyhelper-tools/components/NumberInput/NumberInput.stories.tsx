import { StoryFn } from '@storybook/react';
import { NumberInput, Props } from './NumberInput';

const StoryProps = {
  title: 'Components/TOOLS/NumberInput',
  component: NumberInput,
};

const Template: StoryFn<Props> = (args) => <NumberInput {...args} />;

export const Default = Template.bind({});
Default.args = {
  id: 'id',
  name: 'number input',
  defaultValue: '',
  className: 'rounded-lg',
  onChange: () => {},
};

export default StoryProps;
