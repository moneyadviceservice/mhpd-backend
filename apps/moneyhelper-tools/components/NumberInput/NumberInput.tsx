import {
  NumberFormatValues,
  NumericFormat,
  NumericFormatProps,
} from 'react-number-format';
import { twMerge } from 'tailwind-merge';
import { useId } from 'react';

export interface Props extends NumericFormatProps {
  isFullWidth?: boolean;
  dataTestId?: string;
}

export const NumberInput = ({
  id,
  name,
  defaultValue,
  onChange,
  className,
  isFullWidth = true,
  dataTestId = 'number-input',
  ...rest
}: Props) => {
  const generatedId = useId();
  id = id ?? generatedId;

  const MAX_INPUT_VALUE = 999999999999;

  const minMaxNumberCheck = ({ floatValue }: NumberFormatValues) => {
    if (!floatValue) {
      return true;
    }
    return floatValue <= MAX_INPUT_VALUE && floatValue >= 0;
  };

  return (
    <NumericFormat
      id={id}
      data-testid={dataTestId}
      name={name}
      className={twMerge(
        'tool-field',
        isFullWidth && 'w-full',
        'border',
        'pl-2',
        'ml-0',
        'h-auto',
        'focus:outline-purple-700',
        'focus:shadow-focus-outline',
        className,
      )}
      defaultValue={defaultValue}
      decimalSeparator="."
      decimalScale={3}
      isAllowed={minMaxNumberCheck}
      onChange={onChange}
      {...rest}
    />
  );
};
