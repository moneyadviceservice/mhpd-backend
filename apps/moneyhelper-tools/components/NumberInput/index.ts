export { NumberInput } from './NumberInput';
export type { Props } from './NumberInput';
