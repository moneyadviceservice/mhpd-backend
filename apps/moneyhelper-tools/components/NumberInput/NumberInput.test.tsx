import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { NumberInput } from './NumberInput';

describe('NumberInput component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <NumberInput id="id" name="name" defaultValue="" onChange={() => {}} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with a custom className', () => {
    const { container } = render(
      <NumberInput
        className="rounded-lg"
        id="id"
        name="name"
        defaultValue=""
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with default props', () => {
    const { getByTestId } = render(<NumberInput />);
    const inputElement = getByTestId('number-input');

    expect(inputElement).toBeInTheDocument();
    expect(inputElement.tagName).toBe('INPUT');
    expect(inputElement.getAttribute('type')).toBe('text');
  });

  it('sets id prop correctly', () => {
    const { getByTestId } = render(<NumberInput id="test-id" />);
    const inputElement = getByTestId('number-input');

    expect(inputElement.id).toBe('test-id');
  });

  it('calls onChange handler correctly', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<NumberInput onChange={onChange} />);
    const inputElement = getByTestId('number-input');

    fireEvent.change(inputElement, { target: { value: '123' } });

    const eventObject = onChange.mock.calls[0][0];
    const targetValue = eventObject.target.value;
    expect(targetValue).toBe('123');
  });

  it('applies custom class name', () => {
    const { getByTestId } = render(<NumberInput className="custom-class" />);
    const inputElement = getByTestId('number-input');

    expect(inputElement.classList.contains('custom-class')).toBe(true);
  });
});
