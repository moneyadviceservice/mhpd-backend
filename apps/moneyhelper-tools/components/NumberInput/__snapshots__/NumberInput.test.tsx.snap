// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput component renders correctly 1`] = `
<input
  class="tool-field w-full border pl-2 ml-0 h-auto focus:outline-purple-700 focus:shadow-focus-outline"
  data-testid="number-input"
  id="id"
  inputmode="numeric"
  name="name"
  type="text"
  value=""
/>
`;

exports[`NumberInput component renders correctly with a custom className 1`] = `
<input
  class="tool-field w-full border pl-2 ml-0 h-auto focus:outline-purple-700 focus:shadow-focus-outline rounded-lg"
  data-testid="number-input"
  id="id"
  inputmode="numeric"
  name="name"
  type="text"
  value=""
/>
`;
